### PR TITLE
dnsdist: Fix some issues reported by Thread Sanitizer

### PR DESCRIPTION
**Pull request by rgacogne**
### Short description

None are important issues but should be fixed anyway:
- a possible race when reloading a DoH or DoT certificate, very unlikely on x86_64 but could be an issue on other platforms under huge load ;
- a race in QPSAction and QPSPoolAction could lead to the QPS rate being under-evaluated ;
- a race in TeeAction could lead to under-evaluated metrics.

It would be nice to add a TSAN-enabled build to our CI, using the included suppressions list and passing something like: `TSAN_OPTIONS=halt_on_error=1:suppressions=/path/to/dnsdist-tsan.supp`

### Checklist

I have:
- [x] read the [CONTRIBUTING.md](https://github.com/PowerDNS/pdns/blob/master/CONTRIBUTING.md) document
- [x] compiled this code
- [x] tested this code
- [ ] included documentation (including possible behaviour changes)
- [x] documented the code
- [ ] added or modified regression test(s)
- [x] added or modified unit test(s)

### CODE REVIEW

**Comment by Habbie** [comment]
> It would be nice to add a TSAN-enabled build to our CI, using the included suppressions list and passing something like: `TSAN_OPTIONS=halt_on_error=1:suppressions=/path/to/dnsdist-tsan.supp`

I have not had a look at all - can we combine this with ASAN/UBSAN or would it need to be separate?

**Comment by rgacogne** [comment]
> I have not had a look at all - can we combine this with ASAN/UBSAN or would it need to be separate?

Unfortunately TSAN can't be enabled along with ASAN or UBSAN. I pushed a first attempt at doing a TSAN-enabled dnsdist build and regression tests in CircleCI, but advices are welcome :)

File: .circleci/config.yml [hunk 1]
```diff
@@ -1908,6 +1908,91 @@ jobs:
...
+            CXXFLAGS="-O1 -Werror=vla -Werror=shadow -Wformat=2 -Werror=format-security -Werror=string-plus-int -Wp,-D_GLIBCXX_ASSERTIONS" \
+            ./configure \
+            --enable-option-checking=fatal \
+            --enable-unit-tests \
+            --enable-dnstap \
+            --enable-dnscrypt \
+            --enable-dns-over-tls \
+            --enable-dns-over-https \
+            --enable-systemd \
+            --prefix=/opt/dnsdist \
```
Comment (Habbie) [inline .circleci/config.yml:1970]:
```suggestion
            --prefix=/opt/dnsdist-with-tsan \
```

File: .circleci/config.yml [hunk 2]
```diff
@@ -1939,6 +2024,36 @@ jobs:
...
+            /etc/init.d/snmpd start
+      - run:
+          name: install prometheus tools
+          command: |
+            apt-get -qq --no-install-recommends install prometheus
+      - run:
+          name: Run dnsdist tests
+          workdir: ~/project/regression-tests.dnsdist
+          command: |
+            DNSDISTBIN="/opt/dnsdist/bin/dnsdist-with-tsan" \
```
Comment (Habbie) [inline .circleci/config.yml:2054]:
```suggestion
            DNSDISTBIN="/opt/dnsdist-with-tsan/bin/dnsdist" \
```

**Review by Habbie — CHANGES_REQUESTED** [review]
two diffs suggested - unless I'm totally misunderstanding something!

File: .circleci/config.yml [hunk 3]
```diff
@@ -1908,6 +1908,91 @@ jobs:
...
+          command: make check || (cat test-suite.log; false)
+          working_directory: ~/project/pdns/dnsdistdist
+      - run:
+          name: Install resulting binaries
+          command: make install
+          working_directory: ~/project/pdns/dnsdistdist
+      - persist_to_workspace:
+          root: /opt
+          paths:
+            - dnsdist
```
Comment (Habbie) [inline .circleci/config.yml:1994]: I believe this would overwrite the non-tsan build (or vice versa, i.e., this TSAN PR introduces a race condition ;) )

Comment (rgacogne) [reply .circleci/config.yml:1994]: Ah, I was wondering about that! It's not very clear to me how `persist_to_workspace` works :-/

Comment (Habbie) [reply .circleci/config.yml:1994]: The workspace is per workflow and both the tsan and non-tsan builds are in the `build-and-test-all` workflow. You could make a separate workflow (but I don't really see benefits to that), or take my previous suggestions.

Comment (Habbie) [reply .circleci/config.yml:1994]: Ah, you already did that!

Comment (rgacogne) [reply .circleci/config.yml:1994]: I think I'm starting to see it, thanks! It looks like I also need to change the `persist_to_workplace` path to `dnsdist-with-tsan`, right?

Comment (Habbie) [reply .circleci/config.yml:1994]: yes!

**Review by Habbie — APPROVED** [review]
circleci changes look good now!

**Comment by rgacogne** [comment]
I squashed some commits to clean up the history, will merge as soon as the tests are green again.

**Pull request merged**